include/open62541/plugin/nodestore.h

```diff
@@ -391,12 +388,43 @@ typedef struct UA_LocalizedTextListEntry {
     UA_LocalizedText localizedText;
 } UA_LocalizedTextListEntry;
 
+/* Index of the different permission on the roles array. Every element of the
+ * array carrys the RoleSet with all roles that have the permission. */
+#define UA_ROLEPERMISSIONINDEX_BROWSE 0
+#define UA_ROLEPERMISSIONINDEX_READROLEPERMISSIONS 1
+#define UA_ROLEPERMISSIONINDEX_WRITEATTRIBUTE 2
+#define UA_ROLEPERMISSIONINDEX_WRITEROLEPERMISSIONS 3
+#define UA_ROLEPERMISSIONINDEX_WRITEHISTORIZING 4
+#define UA_ROLEPERMISSIONINDEX_READ 5
+#define UA_ROLEPERMISSIONINDEX_WRITE 6
+#define UA_ROLEPERMISSIONINDEX_READHISTORY 7
+#define UA_ROLEPERMISSIONINDEX_INSERTHISTORY 8
+#define UA_ROLEPERMISSIONINDEX_MODIFYHISTORY 9
+#define UA_ROLEPERMISSIONINDEX_DELETEHISTORY 10
+#define UA_ROLEPERMISSIONINDEX_RECEIVEEVENTS 11
+#define UA_ROLEPERMISSIONINDEX_CALL 12
+#define UA_ROLEPERMISSIONINDEX_ADDREFERENCE 13
+#define UA_ROLEPERMISSIONINDEX_REMOVEREFERENCE 14
+#define UA_ROLEPERMISSIONINDEX_DELETENODE 15
+#define UA_ROLEPERMISSIONINDEX_ADDNODE 16 /* AddNode is part of the
+                                           * DefaultRolePermissions but not
+                                           * included in every Node */
+
+#define UA_ROLEPERMISSIONS_COUNT 16 /* Excluding AddNode */
+
 /* Every Node starts with these attributes */
 struct UA_NodeHead {
     UA_NodeId nodeId;
     UA_NodeClass nodeClass;
     UA_QualifiedName browseName;
 
+    /* The PermissionType defines 17 different RolePermissions. The AddNode role
+     * is global and not defined for each node. For all other 16 PermissionTypes
+     * the node keeps a role-set of all the roles which have this permission.
+     * When a RolePermission is changed manually (added or removed) the changes
+     * propagate recursively to all child nodes (hierarchical reference). */
+    UA_RoleSet rolePermissions[UA_ROLEPERMISSIONS_COUNT];
```

Comment:
This looks like quite some overhead here. Beside that not every node actually has RolePermissions assigned (see comment below) it also poses an upper limit of roles that may be added at runtime.

Wouldn't it be better to keep the RolePermissions outside of the node and accessing them (or only check for access) through an abstraction? That way it would reduce the overhead here and it would also allow to deviate from the standard role mapping which is supported by the OPC UA specification?

Reply:
These are 32byte overhead per node.
16 permission types and 2 byte per type (assuming up to 16 roles).

Reply:
Fair enough, the current size of UA_Node is already at around 400 Bytes so this increase is only about 8%.
And also from what I see, Namespace 0 contains around 5000 nodes. So yes, this overhead seems to be negligible.

Nevertheless I want to point two things out here, that might be worth considering:

* I would expect the amount of permutations for these RolePermissions to be strictly smaller than the overall number of nodes.
  Why? Because first, it seems to be a maintenance burden to assign different role permissions to each node.
  And second, doing this does not make sense at all because of the inter-dependencies between the nodes.
* It does not feel right to have this matrix embedded inside the UA_Node structure. ;-)
  Having a pointer here for example allows to easily detect that no RolePermissions have been assigned (e.g. to fall back to the namespace defaults). And it would also allow to add roles at runtime without any limitations.